### PR TITLE
Cloudmlsm

### PR DESCRIPTION
**Pull request by brandondutra**
updates
--use save model, makes two models (one with and without target columns)
--csv only path for local/cloud. Will support BQ later
--update the dl interface functions

### CODE REVIEW

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 1]
```diff
@@ -100,83 +100,106 @@ def _package_to_staging(staging_package_url):
...
                 top_n=None,
-                layer_sizes=None):
+                layer_sizes=None,
+                learning_rate=0.01,
+                epsilon=0.0005):
   """Train model locally.
   Args:
-    train_file_pattern: train csv file
-    eval_file_pattern: eval csv file
+    train_file_pattern: CsvDataSet
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: Args names don't match args now.

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 2]
```diff
@@ -230,161 +258,153 @@ def local_train(train_file_pattern,
...
+     learning_rate: tf.train.AdamOptimizer's learning rate,
+     epsilon: tf.train.AdamOptimizer's epsilon value.
   """
-  #TODO(brandondutra): allow other flags to be set like batch size/learner rate
+  if len(train_dataset.input_files) != 1 or len(eval_dataset.input_files) != 1:
+    raise ValueError('CsvDataSets must be built with a file pattern, not list '
+                     'of files.')
+
+  if file_io.file_exists(output_dir):
+    raise ValueError('output_dir already exist. Use a new output path.')
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: We don't support incremental training?

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 3]
```diff
@@ -230,161 +258,153 @@ def local_train(train_file_pattern,
...
+                min_eval_frequency=100,
                 top_n=None,
                 layer_sizes=None,
-                project_id=None,
-                job_name=None,
-                scale_tier='STANDARD_1',
-                region=None):
+                learning_rate=0.01,
+                epsilon=0.0005,
+                job_name=None,):
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: extra comma.

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 4]
```diff
@@ -447,20 +471,12 @@ def cloud_predict(model_name, model_version, data, is_target_missing=False):
...
+    data: List of csv strings or a Pandas DataFrame that match the model schema.
 
   Before using this, the model must be created. This can be done by running
   two gcloud commands:
   1) gcloud beta ml models create NAME
-  2) gcloud beta ml models versions create VERSION --model NAME \
+  2) gcloud beta ml versions create VERSION --model NAME \
       --origin gs://BUCKET/training_output_dir/model
   or one datalab magic:
   1) %mlalpha deploy --name=NAME.VERSION \
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: update that command. It should be "ModelVersions('mymodel').deploy('myversion')".

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 5]
```diff
@@ -497,52 +508,70 @@ def cloud_predict(model_name, model_version, data, is_target_missing=False):
...
   """Local batch prediction.
 
   Args:
-    model_dir: local file path to trained model. Usually, this is 
-        training_output_dir/model.
+    training_ouput_dir: The output folder of training.
     prediction_input_file: csv file pattern to a local file.
-    output_dir: local output location to save the results.
+    output_dir: output location to save the results.
+    mode: evaluation or prediction. If evaluation, the input data must contain
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: 'evaluation' or 'prediction'. (Add quotes).

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py [hunk 6]
```diff
@@ -262,40 +140,48 @@ def get_experiment(output_dir):
...
+
+    export_strategy_csv = util.make_export_strategy(
+        train_config=train_config, 
+        args=args, 
+        keep_target=True,
+        assets_extra=additional_assets)
+    export_strategy_json = util.make_export_strategy(
+        train_config=train_config, 
+        args=args, 
+        keep_target=False,
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py]: This is an old name. Renamed  to export_strategy_target, export_strategy_notarget

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 7]
```diff
@@ -41,97 +70,244 @@ def _copy_all(src_files, dest_dir):
...
+      # build maps from ints to the origional categorical strings.
+      string_value = get_vocabulary(args.preprocess_output_dir, target_name)
+      table = tf.contrib.lookup.index_to_string_table_from_tensor(
+          mapping=string_value,
+          default_value='UNKNOWN')
+
+      # Get the label of the input target.
+      if keep_target:
+        input_target_label = table.lookup(input_ops.labels)
+        #input_target_label = tf.contrib.lookup.index_to_string(
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 8]
```diff
@@ -373,8 +552,10 @@ def preprocess_input(features, target, train_config, preprocess_output_dir,
                            'hash_sparse are supported.')
       if map_vocab:
         labels = train_config['vocab_stats'][name]['labels']
-        features[name] = tf.contrib.lookup.string_to_index(features[name],
-                                                           labels)
+        table = tf.contrib.lookup.string_to_index_table_from_tensor(labels)
+        features[name] = table.lookup(features[name])
+        #features[name] = tf.contrib.lookup.string_to_index(features[name],
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py [hunk 9]
```diff
@@ -575,10 +756,10 @@ def merge_metadata(preprocess_output_dir, transforms_file):
   result_dict['target_column'] = schema[0]['name']
   for name, trans_config in transforms.iteritems():
     if trans_config.get('transform', None) == 'target':
-      if name != result_dict['target_column']:
-        raise ValueError('Target from transform file does not correspond to '
-                         'the first column of data')
+      result_dict['target_column'] = name
       break
+  if result_dict['target_column'] is None:
+    raise ValueError('Trarget transform missing form transfroms file.')
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py]: thanks

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 10]
```diff
@@ -100,83 +100,106 @@ def _package_to_staging(staging_package_url):
...
-  args = ['local_preprocess',
-          '--input_file_pattern=%s' % input_file_pattern,
-          '--output_dir=%s' % output_dir,
-          '--schema_file=%s' % schema_file]
+  import datalab.mlalpha as mlalpha
+  if not isinstance(dataset, mlalpha.CsvDataSet):
+    raise ValueError('Only CsvDataSet is supported')
+
+  if len(dataset.input_files) != 1:
+    raise ValueError('CsvDataSet should be built with a file pattern, not a '
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: agree

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 1 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 2 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: Kinda, Lets say training is ran with max_steps=1000. User would have to run again with max_steps=1100 to perform an additional 100 steps (not an additional 1100 steps). This is confusing, so I don't allow it. Once I move away from tf.learn I can do this.

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 3 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: nice, thanks

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 4 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 5 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: done

**Pull request merged**